File: backend/src/services/fairPrice.js
const LEVEL_SCORE = { LOW: 0.25, MEDIUM: 0.5, HIGH: 0.85 };

const SEASON_ADJ = {
  summer: 0.04,
  monsoon: -0.02,
  winter: 0.01,
  harvest: 0,
};

function roundRupee(n) {
  return Math.round(n * 2) / 2;
}

/**
 * Rule-based fair-price band for SIH demo.
 * Weights current mandi price, history, demand, supply, season, lot size.
 * Does not claim ML accuracy.
 */
function recommendFairPrice({
  productName,
  listedPrice,
  quantity = 0,
  market,
}) {
  const current = market?.currentPrice ?? listedPrice ?? 25;
  const historical = market?.historicalAvg ?? current * 0.92;
  const demand = LEVEL_SCORE[market?.demandLevel] ?? 0.5;
  const supply = LEVEL_SCORE[market?.supplyLevel] ?? 0.5;
  const seasonKey = (market?.season || "harvest").toLowerCase();
  const seasonAdj = SEASON_ADJ[seasonKey] ?? 0;

  const blended = current * 0.45 + historical * 0.35 + current * 0.2;
  const demandAdj = (demand - 0.5) * 0.14;
  const supplyAdj = (0.5 - supply) * 0.12;
  const qtyAdj = quantity >= 250 ? -0.04 : quantity >= 100 ? -0.015 : quantity > 0 && quantity < 40 ? 0.025 : 0;

  const midpoint = blended * (1 + demandAdj + supplyAdj + seasonAdj + qtyAdj);
  const min = roundRupee(midpoint * 0.96);
  const max = roundRupee(midpoint * 1.08);
  const recommended = roundRupee(midpoint);

  const reasons = [
    `${productName || "This crop"} mandi/market reference is ₹${current}/kg.`,
    `90-day historical average is ₹${historical}/kg (weight 35%).`,
    `Demand is ${market?.demandLevel || "MEDIUM"} and supply is ${market?.supplyLevel || "MEDIUM"}, so the band is shifted ${demandAdj + supplyAdj >= 0 ? "up" : "down"} slightly.`,
    `Season factor (${market?.season || "harvest"}): ${Math.round(seasonAdj * 100)}%.`,
    quantity
      ? `Listed quantity ${quantity} kg applies a ${qtyAdj >= 0 ? "small-lot premium" : "bulk discount"} of ${Math.round(qtyAdj * 1000) / 10}%.`
      : "Lot size was not provided; no quantity adjustment.",
    "This is a transparent rule engine for the prototype — not a guaranteed market forecast.",
  ];

  let listedAdvice = null;
  if (listedPrice != null) {
    if (listedPrice < min) {
      listedAdvice = `₹${listedPrice}/kg is below the suggested band. Raising toward ₹${min}–₹${max} can improve earnings without matching inflated retail markups.`;
    } else if (listedPrice > max) {
      listedAdvice = `₹${listedPrice}/kg is above the suggested band. Buyers may compare nearby listings and skip this lot.`;
    } else {
      listedAdvice = `₹${listedPrice}/kg sits inside the fair band. It is competitive for a direct farm sale.`;
    }
  }

  return {
    productName,
    marketCurrent: current,
    historicalAvg: historical,
    demandLevel: market?.demandLevel || "MEDIUM",
    supplyLevel: market?.supplyLevel || "MEDIUM",
    season: market?.season || "harvest",
    listedPrice,
    recommended,
    band: { min, max },
    display: `₹${min}–₹${max}/kg`,
    reasons,
    listedAdvice,
    method: "weighted-rules",
  };
}

module.exports = { recommendFairPrice };

File: backend/python/fair_price.py
"""
Optional companion to backend/src/services/fairPrice.js
Same weights, for judges who want to see a Python snippet.
This file is documentation + a CLI helper — the live API uses the Node engine.
"""

LEVEL = {"LOW": 0.25, "MEDIUM": 0.5, "HIGH": 0.85}
SEASON = {"summer": 0.04, "monsoon": -0.02, "winter": 0.01, "harvest": 0}


def recommend(product_name, listed, qty, current, historical, demand="HIGH", supply="MEDIUM", season="harvest"):
    blended = current * 0.45 + historical * 0.35 + current * 0.2
    demand_adj = (LEVEL[demand] - 0.5) * 0.14
    supply_adj = (0.5 - LEVEL[supply]) * 0.12
    season_adj = SEASON.get(season, 0)
    qty_adj = -0.04 if qty >= 250 else (-0.015 if qty >= 100 else (0.025 if 0 < qty < 40 else 0))
    mid = blended * (1 + demand_adj + supply_adj + season_adj + qty_adj)
    return round(mid * 0.96, 1), round(mid, 1), round(mid * 1.08, 1)


if __name__ == "__main__":
    lo, mid, hi = recommend("Tomato", 22, 100, 25, 23)
    print(f"Tomato recommended band ₹{lo}–₹{hi}/kg (mid ₹{mid})")
